### DISCUSSION TITLE
% to Stun and creature summon

### DISCUSSION THREAD

**Discussion by Danimal696** [flareteam/flare-game#907] [Q&A]
Is it possible? and how could I create it?. Taking "Bleed" as a base  is there a way to declare a % for activation?.

Also, is there a way to make summons to autoscale levels without a need to declare more than one?
Ex: "summon rat" so I only need to make a rat.txt insted of Rat_lv1; rat_lv5.txt...

**Comment by dorkster** [comment]
>Is it possible? and how could I create it?. Taking "Bleed" as a base is there a way to declare a % for activation?

When using `post_effect`, the 4th (optional) parameter is used for the activation chance. So if you wanted a 25% chance of stun for 1 second, you would use: `post_effect=stun,1,1s,25`. The value of the second parameter (magnitude) is unimportant here, since stun doesn't use it.

>Also, is there a way to make summons to autoscale levels without a need to declare more than one?

When spawning via a power, you can use the `spawn_level` attribute to scale by either player level or player primary stat:

- `spawn_level=level,1,2`; the enemy will get 1 level for every 2 levels the player has
- `spawn_level=stat,1,2,mental`; the enemy will get 1 level for every 2 points of Mental that they have
- `spawn_level=fixed,2`; the enemy will be level 2

**Reply by Danimal696** [reply]
Thanks for answering; in which way is handled the scaling attribute wise?. Are the declared stats of an enemy raised by an amount depending on player level or a multiplication of it´s base lv?

Ex: player lv:20 = enemy +10lvs
Rat:
stat=hp,5 **x10** => 50
stat=dmg_melee_min,20 **x10** => 200
_or_
stat=hp,5 **+ 10** => 15
stat=dmg_melee_min,20 **+10** => 30

**Reply by dorkster** [reply]
Just like with the player, you would define `stat_per_level` attributes for the summon. Let's look at the [zombie summon](https://github.com/flareteam/flare-game/blob/master/mods/empyrean_campaign/enemies/summon_zombie.txt#L17-L22) from flare-game:

```
stat=accuracy,80
stat_per_level=accuracy,5
```

This means the zombie's accuracy is 80 at level 1, 85 at level 2, 90 at level 3, etc.

**Reply by Danimal696** [reply]
Say, this is really useful; can this be used for common enemies? as in enemies that are not spawned by an event, rather created on map load, having a lv. variable (inputed by hand). 
It would hugely reduce my workload not having to set every single enemy stats by hand by "eye standart", instead just declare a base enemy.txt with it´s declared per lv growth stats.

**Reply by dorkster** [reply]
Unfortunately, the `level` attribute for map enemy groups is only used to determine which enemies are added to the random pool. I think adding a `spawn_level` attribute that's identical to the one we have for Powers would be a good move.

**Reply by Danimal696** [reply]
It would be a total developing changer, I could stop right away trying to balance enemies, slap some excel tables to compare player/monsters growth and get to enemy/map making. 

Please, pretty please, integrate that attribute.

**Reply by dorkster** [reply]
I've added the `spawn_level` attribute to map enemy groups: https://github.com/flareteam/flare-engine/commit/5e520aa00b4324fc6e89dd2c2731ee31bc28620d.

However, not everything about an enemy can scale with their level at the moment. I'm not saying that everything *should* be able to scale with the level, but the most glaring omission is XP. Adding XP scaling is on the TODO list: https://github.com/flareteam/flare-engine/issues/1807

**Reply by Danimal696** [reply]
Thanks! I am aiming for a very simple solution, Xp=3*lv. It follows the same way the stats increase, maybe leave open the base xp value per level for modification, after that you could apply a +% depending on difficulty and round down the result.

What else won´t scale?. Another issue I can see would be loot lists, would it be possible to declare the use of loot list if the monster is in a lv range?
Ex: Rat:
Lv1 to lv5 -> Loot list1.txt
Lv6 to lv10 -> Lootlist2.txt

Which takes me to the item random generation, after messing a bit with scripts for potions and the like, I wonder, couldn´t item generation be handled that way? Having a variable that tells the game to override the loot list (in /engine/) and runs an external script that creates the item based on enemy level (thus living along loot list system). I mean, I was able to make it in php, so why can´t it be reproduced into another language?; I have to admit I have no idea of the feasibility of this or how Flare would save an item created that way but that would be the best way to create a easily modificable item generator that is not hardcoded. If it´s doable I also volunteer for the migration of my generator to whatever Flare script language uses, I´ll just need some initial framework and pointers.

**Reply by dorkster** [reply]
Yes, loot is the other major thing that we might want to scale. I think your solution of level ranges would work.

Random item generation is a whole other can of worms. We make assumptions across the engine that an item can be fully defined if only given its ID. Items would need to be changed into something where the ID is only one component of a bigger object. Here's how I would imagine the item storage part of save data to look like:

```
# Level 2 Cloth Shirt with two bonuses in the 6th inventory slot
[item]
location=inventory_carried,5
base_id=4
level=2
bonus=xp_gain,10
bonus=fire_resist,15
```

You could potentially see other things being added here, like sockets, item degradation, and magical charges.

There's no way I'd introduce scripting languages just for this. I would propose the item generator be part of Flare's C++ code, and take a selection of parameters from a text config file. Just like the rest of the engine. When generating an item, the primary (maybe only?) inputs would be base ID and level. The generator would then work with those given some global parameters, such as:

- Min/Max bonus values per level for each bonus type
- Min/Max number of bonuses that can be attached to an item
- Level variation with lower and upper bounds

**Reply by Danimal696** [reply]
I´m not a fan of item degradation since it adds nothing more to the game than one extra click at shops, but socketing sounds great; being modable with .txt like everything else is also a good move.
I´ll upload the code of the generator I made; it´s not the same language but I´m sure it´s a great base if only because of the naming system I put in place so you don´t have to dig for synonymes. 

_The way I would do it would be something like this (almost what I did):_
-Pass seed Lv of item to creator (maybe the drop % be transfered to creatures or made global?)
-Random it to 75% to 150% > Lv10 "seed" would produce something between lv7-lv15, random rarity as well here (an integer value 1 to 5)
-Random over a list of possible outputs (important it´s an open txt to add new item classes) and pass lv/rarity from previous step
->Potions and similar would have wide lv targets (lv1-10 small, lv11-20 normal, lv21-30 big...), same for my runes, pet fetishes, possible socketable gems... since they don´t really need to be randomized, just dropped
->Selected items class takes lv/rarity, then something like this would take place (not full, the naming and magic properties are made before but takes too much space):

```
*(Randomize magical bonus and name, stores for later use)

[item]
id=<?=$id?> **it´s a counter variable inside a loop**
<?if ($traits >= 5) {?>name=<?=["Legendary","Epic","Godly","Mythical","Ancient","Supreme","Fabled","Eternal","Supernatural","Almighty"][rand(0,9)]?> <?=$adj1?> <?=$name?> of the <?=$adj2?> <?;
	} else if ($traits >= 3) {?>name=<?=$adj1?> <?=$name?> of the <?=$adj2?><?;
	} else if ($traits >= 1) {?>name=<?=$adj1?> <?=$name?> <?;
	} else if ($traits == 0) {?>name=<?=$name?> <?;}
?> 
**Depending on number of magic properties the name is printed, each stat has a different adjectives/affixes/suffixes:
0=> "name=Shortbow"
1=> "name= Hunter shortbow" (+Dexterity)
... Max rarity item name gets a bit of deluxe treatment here**

INCLUDE items/base/weapons/ranged/shortbow.txt 
level=<?=$level?> 
price=<?=rand($level * 100, $level * 100 * ($traits * 0.1 + 1))?>  **Price is a random between (base lv*100) to (base lv*100 + % rarity value), so for a lv10/rarity 5 => "price=(1000 to 1500)" the random gives a nice "dirty" effect to price -> price=1286**
dmg=ranged,<?=rand(1, 1 * $level)?>,<?=rand(1 * $level, 3 * $level)?>  **Randomizing dmg for lower and upper values, 3 is the max growth value of this weapon per Lv**
requires_stat=offense,<?=ceil($level * 5 * 0.7)?> **Needed stats to use item**
requires_stat=physical,<?=ceil($level * 5 * 0.3)?> 

**Anything else needed to print like attack speeds, power to replace or weapon permanent specialty like Mp_regen for mage items,+ crit for daggers...**

<? for ($y = 1; $y <= $traits; $y++){echo "$bonus_array[$y]" . PHP_EOL . "";}?> **loop printer**
<?if ($traits >= 5) {?>quality=unique <?; **rarity printer**
	} else if ($traits >= 3) {?>quality=rare <?;
	} else if ($traits >= 1) {?>quality=magic <?;
	} else if ($traits == 0) {?>quality=common <?;}?> **Actually unneeded line**
<?}?>
```
A Lv10, Unique rarity (5 bonus) created item looks like this (real use of code, just a few of the possible bonus are included, add bonus to code to taste, good for making item classes differents):
```
php.exe 1.php startid=1 endid=1 level=10 traits=5
[item]
id=1
name=Almighty Sturdy Shortbow of the magician
INCLUDE items/base/weapons/ranged/shortbow.txt
level=10
price=1394
dmg=ranged,7,22
requires_stat=offense,35
requires_stat=physical,15
bonus=hp,14
bonus=mental,8
bonus=mental,3
bonus=dmg_ranged_max,7
bonus=dmg_ment_max,5
quality=unique
```
The naming system can be expanded easily but it get ridiculous beyond this point; I updated the bonus on the full list so Hp/Mp... get real boost unlike here which is still a test code. Lv of item could be used to check and print base/xxx.txt since you would know what Lv monsters are into a new act, or recicle old sets like Diablo did when raising difficulty. 
I also decided to ditch the old base system and just left a few general ones, since when items types expands it gets too convoluted.
Ex: items\base\weapons\melee\swords
-Daggers
-Sword
-Sword_heavy
This allows to random a "gfx=xxx" between the different models existing for the same weapon type in a Lv bracket (Ex: Lv1 to 10) and outputting different looking swords with similar stats.

My god what a text wall, sorry for this. Link for full generator in Php:
https://www.dropbox.com/s/8c4or0lcwiijv07/random%20item%20generator%20in%20php.7z?dl=0

**Reply by dorkster** [reply]
>I´m not a fan of item degradation since it adds nothing more to the game than one extra click at shops

I agree wholeheartedly. It was just an example of the type of thing that would be possible.

One thing I'm unsure how to approach properly is item naming. The adjectives would need to be translated separately, which would cause problems for languages where the form of a word changes based on the context around it. I would rather save the headache and forgo randomized item names altogether, instead just using the base item's name.

**Reply by Danimal696** [reply]
_I would rather save the headache and forgo randomized item names altogether, instead just using the base item's name._
Not as nice looking but totally understable, the meat of the generator goes into the item making, traders could fill their stock like this as well. But anyways couldn´t they just modify the generator so the items are just the base name or adjetive/name switch places?.

Maybe having an option to have the simple mode as default and the complex one for the ones who want it.

**Reply by WithinAmnesia** [reply]
I had a sort of dream last night about random magical items and shops in a medieval fantasy setting; the levels were kind of lower like with Flare but the items were almost from a game in the 2030's / a player mod or game / monetized private server / indy game of 5-10 similar to it but each with their own identity / quirks.

How close / far are we in terms of random property and random prefix and affix magical items / a random item generation system like with Diablo II / Path of Exile / vanilla WoW random items like these https://classic.wowhead.com/item=9461/charged-gear and to have them with random inventory in shops working in Flare?

Is there a working / semi-working prototype showcasing how to set up / use random magical prefixes / affixes properties for random magical items in Flare? My memory is a bit foggy / hazy; although I think @Danimal696 said something early about this although I am not sure if it was in tinkering stages / early prototyping. I remember something about another person helping Danimal with random item generation scripts and affixes and prefixes and a dialog discussing about how many prefixes and affixes could reasonably work. With how 2 prefixes and 2 affixes could be a sweet-spot of depth verses complexity. I am unsure 100% on the details so I figured to think out loud / ask :-P.

**Reply by Danimal696** [reply]
You need to install php, but generator code it´s all included into my zip. It´s fully working now, you would only need to copy paste between files to have a working generator (since only a few modifiers are used as testing). I also explained all the steps so it´s easy to adapt to different items using "1.php" as a base.
